Add tool to check use of unstructured logs

adisky opened the pull request:
Signed-off-by: Aditi Sharma <adi.sky17@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
It adds script hack/verify-structured-logging which reports use of unstructured logs. This script will be used to prevent regression after packages have been migrated to use structured logging. 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref #98975

#### Special notes for your reviewer:
This PR only adds the static check tool, scripts needs to be created to use this tool. Currently it only reports the use of unstructured logging functions. It could be improved for other checks like use of format specifiers.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging

```

/sig instrumentation

adisky committed:
Add scripts to use klog/logcheck

Signed-off-by: Aditi Sharma <adi.sky17@gmail.com>

serathius commented:
Overall looks good, can you add tests like for metric stability analysis ?https://github.com/kubernetes/kubernetes/blob/master/test/instrumentation/main_test.go

adisky commented:
> Overall looks good, can you add tests like for metric stability analysis ?https://github.com/kubernetes/kubernetes/blob/master/test/instrumentation/main_test.go

sure, will update with tests.

ehashman commented:
/triage accepted
/priority important-soon
/assign

adisky commented:
It is not working currently for Verbose style logs `klog.V(2).Infof()`, Working on it and tests.

ehashman commented on test/instrumentation/logcheck/testdata/src/k8s.io/klog/v2/klog.go:
```diff
@@ -0,0 +1,182 @@
+limitations under the License.
+*/
+
+// This fake package is created as package golang.org/x/tools/go/analysis/analysistest
+// expects test data dependency to be testdata/src
```
This might be a little yak-shavey, but is there any way to automatically generate this file based on the klog interface so we don't have to manually update this when there are interface changes?

adisky replied:
ya this is problematic trying to figure out.

adisky replied:
I tried but it seems with `analysistest`  we have to do this. If someone can suggest right/feasible approach here i am ready to follow.

ehashman commented on test/instrumentation/logcheck/README.md:
```diff
@@ -0,0 +1,4 @@
+This directory contains tool for checking use of unstructured logs in a package. It is created to prevent regression after packages have been migrated to use structured logs.
+
+To run the tool use `go run main.go analyzer.go <package_name>`   
+`e.g go run main.go analyzer.go $KUBE_ROOT/pkg/kubelet/lifecycle/`
```
At some point we will need to wire this up in `./hack/verify-logger.sh` or something like that but that can be in a follow-up PR.

adisky replied:
will do that in followup PR

ehashman reviewed (commented):
Looking great thus far! I'm unfamiliar with the tool so I'm not yet sure how we'll handle the nested `V(N)` cases but looking forward to your solution.

My test run:

```
ehashman@red-dot:~/src/k8s/test/instrumentation/logcheck$ go run main.go analyzer.go ~/src/k8s/pkg/kubelet/lifecycle/
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/handlers.go:67:4: unstructured logging function "Infof" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/handlers.go:74:4: unstructured logging function "Infof" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/handlers.go:80:3: unstructured logging function "Errorf" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/handlers.go:113:4: unstructured logging function "Errorf" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:65:3: unstructured logging function "Errorf" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:79:3: unstructured logging function "Warningf" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:101:3: unstructured logging function "Warningf" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:113:4: unstructured logging function "Warningf" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:126:4: unstructured logging function "Warningf" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:139:4: unstructured logging function "Infof" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:143:4: unstructured logging function "Infof" should not be used
/home/ehashman/src/k8s/pkg/kubelet/lifecycle/predicate.go:147:4: unstructured logging function "Warningf" should not be used
exit status 3
```

ehashman commented:
/cc @BenTheElder

if you have the chance to provide some early feedback on this static analysis we're writing to prevent structured logging regressions

adisky commented:
> Looking great thus far! I'm unfamiliar with the tool so I'm not yet sure how we'll handle the nested `V(N)` cases but looking forward to your solution.

I updated it for nested case as well. The only problem is with the unit tests and klog file :(

adisky commented:
```
INFO: From Testing //test/instrumentation/logcheck:go_default_test:
==================== Test output for //test/instrumentation/logcheck:go_default_test:
--- FAIL: TestAnalyzer (0.00s)
    analysistest.go:282: go tool not available: exec: "go": executable file not found in $PATH
FAIL
```
for above test failures @nikhita suggested to try the approach from here https://github.com/kubernetes/kubernetes/pull/92212. I did this and the error is resolved but facing new error
```
--- FAIL: TestAnalyzer (0.01s)
    analysistest.go:282: go tool not available: 'go env GOROOT' does not match runtime.GOROOT:
        	go env: /usr/local/go
        	GOROOT: GOROOT
```
some googling took me here https://github.com/bazelbuild/rules_go/issues/2370, It is really difficult to run go tests depends on `analysistest` with bazel :/

BenTheElder commented:
Just from a really quick first skim this imports new packages for non-production-binary usage which leads me to:
- We typically do *not* want tools like this in the main module, because our dependency graph is already 🔥 
   - Does this need to be in k/k at all? or just used here? does it make sense for other subprojects to use?
   - Right now we solve this by importing tools not shipped as part of the release in `hack/tools/go.mod`, perhaps it could live there if it does really need to be in k/k

- Definitely want this to be in a `hack/verify-*` to get it into presubmit etc.

Something adventurous for us to consider (in the future perhaps, to avoid having to rewrite this now) is writing this as a golangci-lint lint:
-  https://golangci-lint.run/contributing/new-linters/#how-to-add-a-private-linter-to-golangci-lint
- @dims has been looking at folding more static analysis lints into `golangci-lint` (in `hack/tools/go.mod`) instead of running them one by one because it performs better (IIRC it can cache loading the go packages instead of loading them once per linter) and leaves us with less scripts to maintain. We do this in some subprojects already and it works well.

adisky commented:
>   * Does this need to be in k/k at all? or just used here? does it make sense for other subprojects to use?

@BenTheElder No it does not needs to be in k/k,  other sub projects may wants to use if they want to move to structured logging.


>   * Right now we solve this by importing tools not shipped as part of the release in `hack/tools/go.mod`, perhaps it could live there if it does really need to be in k/k

Moved to hack/tools.

Do i need to explicitly remove every `BUILD` file generated like done here? https://github.com/kubernetes/kubernetes/blob/master/hack/update-bazel.sh#L116? 

How the tools test will run?
https://github.com/kubernetes/kubernetes/pull/99090/files#diff-f9295648a1cd5b6961f4249f2181128c3f1711222de2d1d2fe8511ec66d6d8a7

ehashman commented:
@adisky can you try running `./hack/update-internal-modules.sh` per the CI suggestion? that might fix this.

We can wire this up with a verify bash script in a separate PR.

BenTheElder commented:
> @BenTheElder No it does not needs to be in k/k, other sub projects may wants to use if they want to move to structured logging.

If you think that's likely I'd recommend moving the implementation of the linter to a k-sigs repo so other projects can import it, and then import it here in hack/tools/go.mod Kubernetes doesn't support importing this repo.

It seems likey to me that we'd want to ask other klog projects to start following this pattern. There's also the possibility that components move out to their own repos as well.

adisky commented:
> If you think that's likely I'd recommend moving the implementation of the linter to a k-sigs repo so other projects can import it, and then import it here in hack/tools/go.mod Kubernetes doesn't support importing this repo.

@ehashman what you think? should we move to separate repo? or for now work from here?

nikhita commented:
> If you think that's likely I'd recommend moving the implementation of the linter to a k-sigs repo

Would it make sense to move this to klog since anyone who uses klog would be interested in the linter?

cc @dims

adisky commented:
@nikhita I thought about that too, if we want to run it as a tool like  [misspell](https://github.com/client9/misspell) then a separate repo will be good (ease of installation).

adisky commented:
/test pull-kubernetes-unit
/test pull-kubernetes-node-e2e

ehashman commented:
> Would it make sense to move this to klog since anyone who uses klog would be interested in the linter?

I'm okay with moving it at some point but we have limited time before code freeze next week and I'd love to get this tool working against the kubelet ASAP, can we ensure this works for CI here, starting off with having this live in `./hack`, and start the migration to a new repo simultaneously?

ehashman commented on hack/update-bazel.sh:
```diff
@@ -114,3 +114,4 @@ fi
 
 # Avoid bazel stuff in tools/ directory
 rm hack/tools/BUILD
+find hack/tools/logcheck/ -name "BUILD" -type f  -delete
```
Should we just run this against everything in `hack/tools` and replace the first command? Is there something that needs it?

adisky replied:
bazel is gone :)

adisky commented:
Updated the PR. The above failures are not related to this PR. I have removed unit tests for the tool as it is not possible to run test(`make test` or k8s-CI expect this) without vendoring dependencies.  
If we do not want to add any dependency to k/k the best bet here is to remove unit test for the tool, we can add them later once we move it some other repo. If it looks to fine to everyone, I will create a PR for the script.

ehashman commented:
@adisky the verify failure and possibly dep failures are both real, you will need to `Run ./hack/update-internal-modules.sh`

IMO let's remove the unit test for now, we can add that back when we migrate it. We need to get this ready to run against the kubelet before EOW.

adisky commented:
@ehashman current failures are because of go 1.16 update in prow, k8s update is following up. All verify scripts are passing at local. See https://github.com/kubernetes/kubernetes/pull/99634

ehashman commented:
I see... if this looks good to SIG Testing for now (cc @BenTheElder) I'd like to get this merged so we can follow this PR up with adding a verify script to prevent directories from regressing back to non-structured logging before code freeze next week.

/lgtm

dims commented:
@adisky please move the following files 3 files to  `https://github.com/kubernetes/klog/` under `tools/logcheck` directory
```
 hack/tools/logcheck/README.md   |  4 ++
 hack/tools/logcheck/logcheck.go | 89 +++++++++++++++++++++++++++++++++
 hack/tools/logcheck/main.go     | 25 +++++++++
```

please add a `go.mod` file in `tools/` directory as well (so we don't mess up the root go.mod). Once that lands, we can update the the following to use the references to `k8s.io/klog/tools/logcheck`
```
 hack/tools/Makefile             |  4 ++
 hack/tools/go.mod               |  1 +
 hack/tools/tools.go             |  2 +
```

thanks! sorry for the delay

adisky commented:
@dims Added a PR in klog https://github.com/kubernetes/klog/pull/213

BenTheElder commented:
Thanks all! Catching up on this now. Commented in https://github.com/kubernetes/klog/pull/213

dims commented:
@adisky i’ve cut a klog with the linter https://github.com/kubernetes/klog/releases/tag/v2.6.0 please go ahead and use that.

ehashman commented:
Hell yes!

Looks like verify and dependencies have some legit CI failures... I might try to pick this up if @adisky is done for the day.

ehashman commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,38 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.#!/usr/bin/env bash
```
```suggestion
# limitations under the License.
```

There's an extra shebang here causing the boilerplate verification to fail.

ehashman commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,38 @@
+for package in $migrated_packages
+do
+	echo "verifying structured logging for $package"
+	find $KUBE_ROOT/$package -type d -print | xargs logcheck || continue
```
Shellcheck is mad about this line, it suggests using `-print0` but xargs doesn't love that. Also want quotes around the path for `find`.

BenTheElder replied:
`find ... -print0 | xargs -0 ...`, or `find ... -exec` (no xargs) should work.

ehashman reviewed (commented):
Not quite sure what's going on with the internal deps but `./hack/update-internal-dependencies.sh` did make some changes for me.

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,38 @@
+  GO111MODULE=on go install k8s.io/klog/hack/tools/logcheck
+popd >/dev/null
+
+for package in $migrated_packages
```
I suspect in the future we'll want some concurrency in this, but given we only have 2 packages for now, perhaps just a TODO?

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,38 @@
+source "${KUBE_ROOT}/hack/lib/init.sh"
+
+migrated_files="${KUBE_ROOT}/hack/.structured_logging"
+migrated_packages=$(sed "s| | -e |g" "${migrated_files}")
```
more verbose:
```suggestion
migrated_packages=()
while IFS='' read -r line; do
  migrated_packages+=("./$line")
done < <(cat "${migrated_files}")
```
but sed + capture to array here is a bit of a hack.
instead of `./$line` we could put ./ in the file.

adisky replied:
Done.

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,38 @@
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+export KUBE_ROOT
```
there's no reason to export this?
we usually do not.

adisky replied:
Done.

BenTheElder commented on hack/.structured_logging:
```diff
@@ -0,0 +1,2 @@
+pkg/kubelet/dockershim
```
Usually this file tracks all of the failing packages that _should_ be migrated / fixed but are not yet, so we can reduce the exception list over time and eventually not allow exceptions.

I don't consider this a blocker, but just a suggestion to bring this inline with `hack/verify-*`, so far I think it has worked well. It also discourages brand new packages from failing a linter because then you also have to add them to the known failures file in that PR and someone has to approve that instead of having it fixed up front (like the shellcheck errors in an earlier iteration of this PR).

adisky replied:
I saw how other scripts are working, but currently all k/k packages are not migrated to structured logging. It will fail for every package, the idea here is when a package is completely migrated we add to this file and there is currently one package.

adisky replied:
so should we add a list of all packages here as failures?

ehashman replied:
I think this should be an allow-list/list of things that are migrated for now, because the goal is to set this file to pkg/kubelet and cmd/kubelet by 1.21 release freeze. We are doing a slow migration and want to prevent regressions, and because structured logging is still in beta, we don't want to prevent new files from regressing quite yet. When we're ready to make the GA switch, I think we can flip this.

BenTheElder replied:
+1 re: https://github.com/kubernetes/kubernetes/pull/99090#discussion_r587620282 (@ehashman comment just above this one)

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,38 @@
+for package in $migrated_packages
+do
+	echo "verifying structured logging for $package"
+	find "$KUBE_ROOT/$package" -type d -print0 | xargs -0 logcheck || continue
```
I don't think we need the find call here at all actually, in the future we're going to need to use k8s.io/kubernetes/$pkg or ./$pkg anyhow.

we can just do `logcheck $pkg` where $pkg is `./pkg/foo`

adisky replied:
currently linter is not working recursively and it expects complete path.

BenTheElder replied:
oh I see, the file is only listing top level directories to enforce?

I think we should have a comment somewhere explaining that intent, usually we list individual packages that are excluded, or individual files.

adisky replied:
yes, right it is only listing top level directories. Sure will add a comment here.

adisky replied:
Updated .structured_logging to add all directories, not going against pattern here :). Previous approach was confusing.

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,38 @@
+export KUBE_ROOT
+source "${KUBE_ROOT}/hack/lib/init.sh"
+
+migrated_files="${KUBE_ROOT}/hack/.structured_logging"
```
nit: name is misleading. this file lists packages, not files.
```suggestion
migrated_packages_file="${KUBE_ROOT}/hack/.structured_logging"
```

BenTheElder reviewed (commented):
couple suggestions / nits.

more high level:

Usually the linter verify script is a little more complex, but such that:
- ensure checked-in list file is sorted (minimize merge conflicts / arbitrary contents), error if not
- usually we check all packages and require that packages that are not failing be removed from being considered failing (hack/.staticcheck_failures etc.) so they will not fail again
- if a package is failing but is not in a known failure list, we require it to be added (instead of a list of known passing, to discourage adding new packages)

I understand why we might instead want to consider only certain packages at all (it's faster, simpler to implement, and maybe we're not requiring all packages to do this just yet), but perhaps consider matching the approach in the other hack/verify-* linters.

see: hack/verify-staticcheck.sh (a lot of the other ones no longer have a known failures file because we've fixed ~everything finally)

BenTheElder commented:
Thank you for your patience working through these comments, I've been distracted by the go 1.16 issues but will be keeping an eye out to help get this in by code freeze ideally.

FWIW as well: I think we should have plenty of support to merge after code freeze if this misses it somehow, because this only alters _tests_ not the release binaries, and test freeze is not until march 24th https://www.kubernetes.dev/resources/release/#tldr

.. but I think we can get this in before code freeze anyhow 😅

adisky commented:
> usually we check all packages and require that packages that are not failing be removed from being considered failing (hack/.staticcheck_failures etc.) so they will not fail again
if a package is failing but is not in a known failure list, we require it to be added (instead of a list of known passing, to discourage adding new packages)
I understand why we might instead want to consider only certain packages at all (it's faster, simpler to implement, and maybe we're not requiring all packages to do this just yet), but perhaps consider matching the approach in the other hack/verify-* linters.

Thank you @BenTheElder for the review <3. Addressed all comments
except following the pattern of failure files :). As currently we have very few packages migrated to structured logging. We are only targeting for /pkg/kubelet this cycle. If we follow the failure style saving we need to add all the packages in the file.  I am not aware of the caveats of this approach if you still suggest to for failure approach I'll update the script.

ehashman commented:
> FWIW as well: I think we should have plenty of support to merge after code freeze if this misses it somehow, because this only alters _tests_ not the release binaries, and test freeze is not until march 24th https://www.kubernetes.dev/resources/release/#tldr
> 
> .. but I think we can get this in before code freeze anyhow sweat_smile

The hope is to get this merged and included in CI before code freeze because it will validate that we've fully migrated `pkg/kubelet` and `cmd/kubelet` to structured logging, which is an acceptance criteria for structured logging graduation to beta.

ehashman commented on hack/.structured_logging:
```diff
@@ -0,0 +1,17 @@
+pkg/kubelet/dockershim/
+pkg/kubelet/dockershim/cm
+pkg/kubelet/dockershim/libdocker
+pkg/kubelet/dockershim/libdocker/testing
+pkg/kubelet/dockershim/metrics
+pkg/kubelet/dockershim/network
+pkg/kubelet/dockershim/network/cni
+pkg/kubelet/dockershim/network/cni/testing
+pkg/kubelet/dockershim/network/hairpin
+pkg/kubelet/dockershim/network/hostport
+pkg/kubelet/dockershim/network/kubenet
+pkg/kubelet/dockershim/network/metrics
+pkg/kubelet/dockershim/network/testing
+pkg/kubelet/dockershim/remote
+pkg/kubelet/server/metrics
+pkg/kubelet/server/stats
+pkg/kubelet/server/stats/testing
```
Nit: is this smart enough to do a reduced list of

```suggestion
pkg/kubelet/dockershim
pkg/kubelet/server
```

or are there still missing files in those directories?

adisky replied:
nope, see the discussion  https://github.com/kubernetes/kubernetes/pull/99090#discussion_r587043479
I included only top level directories initially `pkg/kubelet/dockershim`, `pkg/kubelet/server` but that can cause confusion as which level directory we should add/we shouldn't and also @BenTheElder suggested we follow this for rest of the scripts.

ehashman replied:
edit: pkg/kubelet/server/server.go isn't migrated so we'd still need to keep the last three

I think this is much more difficult to grok on a dir by dir basis like this, particularly when we've also included all of the subdirectories. But it seems that the script doesn't actually work recursively on the subdirectories, so this would be required. So this is fine for now.

ehashman reviewed (commented):
I think at this point the last thing remaining is to wire this into verify.sh?

adisky commented:
> I think at this point the last thing remaining is to wire this into verify.sh?

I added a verify script, is something else i am missing?
https://github.com/kubernetes/kubernetes/blob/5922c90e354e14fd38b4542ea087a7bf4606b8db/hack/verify-structured-logging.sh

ehashman commented:
Yes, in order for this to run as part of `make verify` I believe https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/verify.sh needs to be updated.

adisky commented:
It is running 🤔 
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99090/pull-kubernetes-verify/1367363075899920384#1:build-log.txt%3A2135

BenTheElder commented:
Verify locates all hack/verify*.sh and runs them except for specific exclusions. Shouldn't need to change anything there.

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,45 @@
+
+# TODO: Improve concurrancy here
+for package in "${migrated_packages[@]}"; do
+  logcheck "$KUBE_ROOT/$package" || continue
```
why is this `|| continue` ? there's nothing else in the loop so we're not continuing past anything but doing `|| continue` will mask the error if there is a failing package?

BenTheElder replied:
```sh
ret=0
for package in "${migrated_packages[@]}"; do
  logcheck "$KUBE_ROOT/$package" || ret=$?
done
exit $ret
```
(and possibly check ret and print a message)

ehashman replied:
Is the ret stuff necessary? It will exit on that return code without this. (Maybe shellcheck was complaining?)

BenTheElder replied:
we could drop the `||continue` and it will exit on the first one, but IMHO it's a better experience to log all errors and then exit failure (so you can fix all of them instead of only seeing the first one we hit until you fix that), using the ret thing lets us retain if any of them failed while running all of them.

adisky replied:
Done.

BenTheElder commented:
we indeed ran it in 13s, https://github.com/kubernetes/kubernetes/pull/99090#discussion_r587742361 is my main remaining concern (it looks like this script can never exit failure currently?)

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,45 @@
+
+# Install logcheck
+pushd "${KUBE_ROOT}/hack/tools" >/dev/null
+  GO111MODULE=on go install k8s.io/klog/hack/tools/logcheck
```
this may also need https://github.com/kubernetes/kubernetes/blob/792b4b1e32394a13c27741f71f14dd134d663a5f/hack/verify-govet-levee.sh#L30-L31

we don't want to install to the global GOBIN

BenTheElder replied:
also needs a call to `kube::golang::setup_env` earlier I think (which will also validate the required go version if running this script by itself).

ehashman replied:
Yes, the setup_env was required for me to run this locally.

adisky replied:
Using `kube::golang::setup_env` not working for me locally,  added `kube::golang::verify_go_version`, also I have checked other linting scripts they are not calling `setup_env`.
https://github.com/kubernetes/kubernetes/blob/master/hack/verify-staticcheck.sh

BenTheElder replied:
`kube::golang::setup_env` is pretty standard, this is a bug in verify-staticcheck if anything https://cs.k8s.io/?q=kube%3A%3Agolang%3A%3Asetup_env&i=nope&files=&excludeFiles=&repos=

adisky commented:
/test pull-kubernetes-unit

ehashman commented:
/retest

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,64 @@
+done < <(cat "${migrated_packages_file}")
+
+
+# TODO: Improve concurrancy here
```
nit: seems spelling lint missed this, `s/concurrency/concurrancy/` (not going to block on this during code freeze of all times!)

BenTheElder replied:
Also hard data: this is taking 13s in CI which is fast enough for now (though nearly as slow as `gofmt` parsing the entire repo). The overall job takes 35m, but every bit counts (most of these tools are not easily made any faster, and we have removed some checks out of the default `make verify` to avoid excessive time creep).
Will investigate in follow-ups.

BenTheElder commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,64 @@
+source "${KUBE_ROOT}/hack/lib/init.sh"
+source "${KUBE_ROOT}/hack/lib/util.sh"
+
+kube::golang::verify_go_version
```
because this operates on the source tree `kube::golang::setup_env` is needed to work for developers that do not have Kubernetes checked out to a GOPATH. we can do that in a follow-up though.

BenTheElder reviewed (approved):
/lgtm
/approve
This is sufficient to iterate on, thank you.
/hold
One remaining check to consider: normally before merging a new lint like this we want to see that it will in fact fail on something, so we typically push a temporary commit to show this (e.g. adding a bad change in a covered file), then revert the change (giving CI runs reviewers can see).

ehashman commented on hack/verify-structured-logging.sh:
```diff
@@ -0,0 +1,64 @@
+# TODO: Improve concurrancy here
+ret=0
+for package in "${migrated_packages[@]}"; do
+  logcheck "$KUBE_ROOT/$package" || ret=$?
```
Not sure if this is the best way to do this (because it will involve building twice) but I was able to fix https://github.com/kubernetes/klog/issues/215 with the following patch:

```suggestion
  GOOS=linux logcheck "$KUBE_ROOT/$package" || ret=$?
  GOOS=windows logcheck "$KUBE_ROOT/$package" || ret=$?
```

ehashman commented:
/hold cancel

I've confirmed locally this fails against files not in the list. I'm going to follow up with a PR to update migrated files and update the guidance for structured logging PRs to add to this file.

BenTheElder commented:
/retest
🤨
